File: 0450-delete-node-in-a-bst/0450-delete-node-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* deleteNode(TreeNode* root, int key) {
        if (root == nullptr){
            return root;
        }

        auto pos = findKey(root, key);
        if(pos.second == nullptr){
            return root;
        }
        
        auto parent = pos.first;
        auto node = pos.second;
        bool left = parent && parent->left == node ;

        if(!node->left && !node->right){
            if (!parent){
                return nullptr;
            }

            if (left){
                parent->left = nullptr;
            }
            else{
                parent->right = nullptr;
            }
            return root;
        }

        else if (!node->left  || !node->right ){
            auto remain_node = node->left ? node->left : node->right;
            if (!parent){
                return remain_node;
            }
        
            if (left) {
                parent->left = remain_node;
            }
            else{
                parent->right = remain_node;
            }
            return root;
        }

        else{
            auto largest_left_subtree = findLargest(node->left);
            auto replace_val = largest_left_subtree->val;

            deleteNode(node, replace_val);

            node->val = replace_val;
            return root;
        }
    }

    TreeNode* findLargest(TreeNode* root){
        if(root->right == nullptr){
            return root;
        }
        return findLargest(root->right);
    }

    pair<TreeNode*, TreeNode*> findKey(TreeNode* root, int key){
        if (root == nullptr){
            return make_pair(nullptr, nullptr);
        }

        if (root->val == key){
            return make_pair(nullptr, root);
        }

        if (root->left != nullptr && root->left->val == key){
            return make_pair(root, root->left);
        }

        else if (root->right != nullptr && root->right->val == key){
            return make_pair(root, root->right);
        }
        
        auto left_result = findKey(root->left, key);
        if (left_result.second != nullptr){
            return left_result;
        }

        return findKey(root->right, key);
    }
};

File: 0450-delete-node-in-a-bst/0450-delete-node-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
        if root is None:
            return None

        if root.val > key:
            root.left = self.deleteNode(root.left, key)
        
        elif root.val < key:
            root.right = self.deleteNode(root.right, key)
        
        else:
            if root.left is None:
                return root.right
            elif root.right is None:
                return root.left
            
            max_in_left_tree = self.find_max(root.left)

            root.val = max_in_left_tree.val

            root.left = self.deleteNode(root.left, max_in_left_tree.val)

        return root

    
    def find_max(self, root):
        while root.right:
            root = root.right

        return root


        